### PR TITLE
Create Bascis Module

### PR DESCRIPTION
**Pull request by philschulz**
This pull request depends on the earlier one. Please only merge this after you have merged the other one.

This should cover everything up to and including the fourth bullet point in modules_outline. Please feel free to comment and suggest changes.

### CODE REVIEW

**Comment by philschulz** [comment]
I now squashed most commits so that you should see most of my changes in one diff.

**Comment by philschulz** [comment]
It's probably best to directly look at the slides, though.

**Comment by philschulz** [comment]
Sorry, I had to rebase locally to resolve the conflicts. The content has not changed, though.

File: modules/M1_Basics/M1_Basics.tex [hunk 1]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\begin{frame}{Factorial HMMs}
+FHMMs have several Markov chains over latent variables.
+\begin{itemize}
+\item $ M $ Markov chains over latent variables.
+\item $ L $ outcomes per latent variable.
+\item Sequence of length $ T $.
+\item Complexity of inference: $ \mathcal{O}(L^{2M}T) $.
+\end{itemize}
+\pause
+This model is \textbf{intractable} because of the exponential dependence on the number of hidden Markov chains.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: Perhaps box this

```latex
\begin{alertblock}{Intractable}
Exponential dependency on the number of hidden Markov chains
\end{alertblock}
```

File: modules/M1_Basics/M1_Basics.tex [hunk 2]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\end{frame}
+
+\begin{frame}{Latent Dirichlet Allocation}
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: This slide is redundant with the first frame of the next slide.

File: modules/M1_Basics/M1_Basics.tex [hunk 3]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+
+\begin{frame}{The Goal}
+Assume $ p(z|x) $ is intractable.
+\pause
+\begin{block}{Idea}
+Let's approximate it by an auxiliary distribution $ q(z) $ that is tractable!
+\end{block}
+\pause
+\begin{block}{Requirement}
+Make $ q(z) $ as close as possible to $ p(z|x) $ to obtain a faithful approximation.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: Use the word *choose* here somewhere to stress the idea that we are searching for a good $q$

File: modules/M1_Basics/M1_Basics.tex [hunk 4]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\begin{itemize}
+\item $ \KL{q(z)}{p(z|x)} = \int q(z) \loga{\frac{q(z)}{p(z|x)}} dz $ (continuous)
+\item $ \KL{q(z)}{p(z|x)} = \sum_{z} q(z) \loga{\frac{q(z)}{p(z|x)}} $ (discrete)
+\end{itemize}
+\end{frame}
+
+\begin{frame}{Recap KL divergence}
+\begin{block}{Properties}
+\begin{itemize}
+\item $ \KL{q(z)}{p(z|x)} \geq 0 $ with equality iff $ q(z) = p(z|x) $.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: break line before *with equality*

File: modules/M1_Basics/M1_Basics.tex [hunk 5]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\pause
+\item $ \KL{q(z)}{p(z|x)} = \infty $ if $ \exists z $ s.t. $ p(z|x) = 0 $ and $ q(z) > 0 $.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: break line before *if*

File: modules/M1_Basics/M1_Basics.tex [hunk 6]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\begin{block}{Idea}
+Let's approximate it by an auxiliary distribution $ q(z) $ that is tractable!
+\end{block}
+\pause
+\begin{block}{Requirement}
+Make $ q(z) $ as close as possible to $ p(z|x) $ to obtain a faithful approximation.
+\end{block}
+\pause
+\begin{block}{Implementation}
+Minimize $ \KL{q(z)}{p(z|x)} $.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: * Define KL also as an expectation: \mathbb E_{q(Z)} \left[ \log \frac{q(Z)}{p(Z|X)} \right]
* Do we write KL(q(z) || p(z|x)) or KL(q(Z)||p(Z|X))?

File: modules/M1_Basics/M1_Basics.tex [hunk 7]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\begin{frame}{Recap KL divergence}
+\begin{block}{Properties}
+\begin{itemize}
+\item $ \KL{q(z)}{p(z|x)} \geq 0 $ with equality iff $ q(z) = p(z|x) $.
+\pause
+\item $ \KL{q(z)}{p(z|x)} = \infty $ if $ \exists z $ s.t. $ p(z|x) = 0 $ and $ q(z) > 0 $.
+\pause
+\item In general $ \KL{q(z)}{p(z|x)} \not = \KL{p(z|x)}{q(z)} $.
+\pause
+\item $ -\KL{q(z)}{p(z|x)} = \int q(z) \frac{p(z|x)}{q(z)} dz \leq 0 $.
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: You are missing a \log here

File: modules/M1_Basics/M1_Basics.tex [hunk 8]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+\end{itemize}
+\end{block}
+\end{frame}
+
+\subsection{Deriving VI with Jensen's Inequality}
+
+\begin{frame}{VI derivation I}
+\begin{small}
+\begin{align*}
+\loga{p(x)} &= \int \loga{p(x,z)} dz \\
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: The brackets in \log p(x,z) are disturbing.

Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: Use $\mathrm dz$ (it's useful to define a macro that takes the integrand, the variable being integrated, and optionally the support

How is it that $\log p(x) = \int \log p(x,z) dz$? Maybe you meant $\log p(x) = \log \left( \int p(x, z) \mathrm dz \right)$ ?

File: modules/M1_Basics/M1_Basics.tex [hunk 9]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+&= \int \loga{\alert{q(z)}\frac{p(x,z)}{\alert{q(z)}}} dz \\
+&\geq \int \alert{q(z)} \loga{\frac{p(x,z)}{\alert{q(z)}}} dz \\
+& = \int \alert{q(z)} \loga{\frac{p(z|x)p(x)}{\alert{q(z)}}} dz \\
+&= \int \alert{q(z)} \loga{\frac{p(z|x)}{\alert{q(z)}}} dz + \loga{p(x)}
+\end{align*}
+\end{small}
+\end{frame}
+
+\begin{frame}{VI derivation I}
+Notice that
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: Some slides (like this one) read like lecture notes. We can write up some lecture notes ;)

File: modules/M1_Basics/M1_Basics.tex (hunk 1 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Done.

File: modules/M1_Basics/M1_Basics.tex (hunk 2 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Good catch! Removed it.

File: modules/M1_Basics/M1_Basics.tex (hunk 3 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Done.

File: modules/M1_Basics/M1_Basics.tex [hunk 10]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
 Again, approximate inference is needed.
 \end{frame}
 
-\end{document}
\ No newline at end of file
+\section{Examples}
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: It won't let me comment on the first few slides. So here are some notes

# General

* we are missing slide numbers
* the blue on red links in the tree-structure of the talk (top bar) is terribly ugly
* I tweaked and improved the bayesnet tikz code (for example, the plate spacing is much better), I will upload a file after this has been merged (I will create an issue for that).

# Slide by slide

## Likelihood and Prior

Let's use boxes as in
```latex
\begin{block}{Hyperparameters}
The prior may depend on a deterministic quantity $\alpha$
\end{block}
```

# Bayes' rule

* Do we need 's? I never see it when people say *Bayes rule* but I am fine with it. Second, Bayes is a proper name, because of that I think it gets the *s* even though it ends in *s*, as in *Bayes's*.
* We don't need to say *Bayes rule asserts that*. It sounds too much like a script.


# Basic Problem

* Recap the role of the evidence in the posterior formula
* Keep the word *inference* for later (assume some people don't even known what inference is, a lot of people use the word inference referring to a prediction of some kind, as in *decoding* with a Viterbi algorithm for some structured prediction problem they know, for example). Let's disambiguate the word inference in this talk.
* Dedicate a slide to the word *Inference* before your slide on *Bayesian inference*

# Bayesian Inference

* no need for *Under the Bayesian view,*.
* line break before *(supervised learning)*
* no need for *This is because*
* do not talk about *approximate inference* just yet

Then add a slide on *Approximate inference* with a remark about its relevance for both frequentist and Bayesian view.

File: modules/M1_Basics/M1_Basics.tex (hunk 6 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Done. Not sure about the supplementary. Shouldn't be more than one page. The slides should be self-contained anyway.

Good question about the notation. I would go with lower-case as this seems to be more standard (I think I also do the same for entropy).

File: modules/M1_Basics/M1_Basics.tex (hunk 4 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Done

File: modules/M1_Basics/M1_Basics.tex (hunk 5 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Done

File: modules/M1_Basics/M1_Basics.tex (hunk 7 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Good catch. I took out the integral. Don' t know why I even had it there.

File: modules/M1_Basics/M1_Basics.tex (hunk 8 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: It's not \log, it's \loga. Comes from vimacros and makes sure that the brackets are the right size.

File: modules/M1_Basics/M1_Basics.tex (hunk 6 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Btw. there is the macro \E{} in vimacros that does exactly what you want for the expectation. For consistency, we should make sure to always use our own macros (in case we wanna change something later).

File: modules/M1_Basics/M1_Basics.tex (hunk 8 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: I defined a macro \intl that takes the integrand as first arg, the integration variable as second arg and the domain of integration as optional arg. Use this from now on. The macro also uses \mathrm.
(TODO: replace integrals everywhere)

I switched everything to \intl here.

File: modules/M1_Basics/M1_Basics.tex (hunk 9 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Well, if you have the time ;) In general I don't like lecture notes because a) they are not necessary for a live tutorial and b) we would be competing with existing lecture notes/written tutorials like Blei's. Why redo something that already exists and is actually really good?

I do see that we should shorten the text here, though. Do you like the new version better?

File: modules/M1_Basics/M1_Basics.tex [hunk 11]
```diff
@@ -93,4 +95,511 @@ \section{Generative Models}
...
+As before, we have derived a lower bound on the log-evidence. This \textbf{evidence lower bound}
+or \textbf{ELBO} is our optimisation objective.
+\begin{block}{ELBO}
+\begin{equation*}
+\underset{q(z)}{\max}~\E[q(z)]{\loga{p(x,z)}} + \Ent{q(z)}
+\end{equation*}
+\end{block}
+\end{frame}
+
+\subsection{Relationship to EM}
```
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Right, I actually didn't focus on the optimization part at all so far. Would you mind adding a slide or two after this is merged?

File: modules/M1_Basics/M1_Basics.tex (hunk 8 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: True about the integral outside the log. Good that you spotted that!

File: modules/M1_Basics/M1_Basics.tex (hunk 10 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Oh fuck, I did not catch your latest comments. Give me a minute here. You can already have a look at my reply. Apart from addressing your comments in the latest push, I also fixed the inference network for LDA everywhere (an arrow from beta to z was missing).

Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Cool, thanks for improving bayesnet. You can already open a separate pull request as there should be no conflict with this one. I didn't touch bayesnet.

For slide numbers and the tree: that's default in Montpellier but you can tweak it if you like.

Using Blocks is good but I also wouldn't overuse them. We should have a convention on them. When would you use them (I basically just used them to have intermediate headings on slides).

About Bayes rule: I find Bayes's ugly but any other variant works for me.

Extra slides in the next commit/PR please.

I will edit (shorten the text) in the final review round of this PR.

File: modules/M1_Basics/M1_Basics.tex (hunk 11 above)
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: I'll do that.

File: modules/M1_Basics/M1_Basics.tex [hunk 12]
```diff
@@ -308,21 +289,22 @@ \section{Variational Inference}
 The Kullback-Leibler divergence (or relative entropy) measures the divergence of a distribution $ q $ from 
 a distribution $ q $. 
 \begin{itemize}
-\item $ \KL{q(z)}{p(z|x)} = \int q(z) \loga{\frac{q(z)}{p(z|x)}} dz $ (continuous)
+\item $ \KL{q(z)}{p(z|x)} = \intl{ q(z) \loga{\frac{q(z)}{p(z|x)}}}{z} $ (continuous)
 \item $ \KL{q(z)}{p(z|x)} = \sum_{z} q(z) \loga{\frac{q(z)}{p(z|x)}} $ (discrete)
+\item $ \KL{q(z)}{p(z|x)} = \E[q(z)]{\loga{\frac{q(z)}{p(z|x)}}} $ (both)
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: E_{q(z)}[f(z)] is ugly, we should use rvs E_{q(Z)}[f(Z)] or E_{q_Z}[f(Z)].

File: modules/M1_Basics/M1_Basics.tex [hunk 13]
```diff
@@ -332,25 +314,23 @@ \subsection{Deriving VI with Jensen's Inequality}
...
+&= \intl{ \alert{q(z)} \loga{\frac{p(z|x)}{\alert{q(z)}}}} {z} + \loga{p(x)}
 \end{align*}
 \end{small}
 \end{frame}
 
 \begin{frame}{VI derivation I}
-Notice that 
 \begin{small}
 \begin{align*}
 \int q(z) \loga{\frac{p(z|x)}{q(z)}} dz + \loga{p(x)} \\
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: \mathrm dz

File: modules/M1_Basics/M1_Basics.tex [hunk 14]
```diff
@@ -332,25 +314,23 @@ \subsection{Deriving VI with Jensen's Inequality}
 \begin{frame}{VI derivation I}
 \begin{small}
 \begin{align*}
-\loga{p(x)} &= \int \loga{p(x,z)} dz \\
-&= \int \loga{\alert{q(z)}\frac{p(x,z)}{\alert{q(z)}}} dz \\
-&\geq \int \alert{q(z)} \loga{\frac{p(x,z)}{\alert{q(z)}}} dz \\
-& = \int \alert{q(z)} \loga{\frac{p(z|x)p(x)}{\alert{q(z)}}} dz \\
-&= \int \alert{q(z)} \loga{\frac{p(z|x)}{\alert{q(z)}}} dz + \loga{p(x)}
+\loga{p(x)} &= \loga{\intl{p(x,z)}{z}} \\
```
Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: log(p(x)) is ugly!

Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Oh, now I see what you mean. Yeah, true, I guess it's fine to just use the standard log here.

File: modules/M1_Basics/M1_Basics.tex (hunk 13 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Right, I changed \int to \intl everywhere. Should be fine now.

File: modules/M1_Basics/M1_Basics.tex (hunk 12 above)
Comment (philschulz) [inline modules/M1_Basics/M1_Basics.tex]: Hmm, alright. If we do this, we have to do it everywhere though (expectation, KL, entropy). I think this would be a separate PR.

Comment (wilkeraziz) [inline modules/M1_Basics/M1_Basics.tex]: Yep, I think we should do it (in a separate PR then).

**Comment by wilkeraziz** [comment]
Can you merge?

**Comment by philschulz** [comment]
Merged and deleted branch :)

Thanks for the reviews! 👍
